src/validation.cpp

```diff
@@ -3808,6 +3808,29 @@ bool HasValidProofOfWork(const std::vector<CBlockHeader>& headers, const Consens
             [&](const auto& header) { return CheckProofOfWork(header.GetHash(), header.nBits, consensusParams);});
 }
 
+bool IsBlockMutated(const CBlock& block, bool check_witness_root)
+{
+    BlockValidationState state;
+    if (!CheckMerkleRoot(block, state)) {
+        LogDebug(BCLog::VALIDATION, "%s: %s\n", __func__, state.ToString());
+        return true;
+    }
+
+    if (block.vtx.empty() || !block.vtx[0]->IsCoinBase()) {
+        return false;
+    }
+
+    if (check_witness_root) {
+        const auto valid_opt = CheckWitnessCommitment(block, state);
```

Comment:
nit c62eea853e5c5f390d58c5ad9c0ac9a688465abc
```suggestion
        const auto valid_opt{CheckWitnessCommitment(block, state)};
```